server/src/main/java/org/elasticsearch/common/util/LazyInitializable.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.elasticsearch.common.CheckedSupplier;
+
+import java.util.Objects;
+import java.util.function.Consumer;
+
+/**
+ * Encapsulates a {@link CheckedSupplier} which is lazily invoked once on the
+ * first call to {@code #getOrCompute()}. The value which the
+ * <code>supplier</code> returns is memorized and will be served until
+ * {@code #reset()} is called. Each value returned by {@code #getOrCompute()},
+ * newly minted or cached, will be passed to the <code>primer</code>
+ * {@link Consumer}. On {@code #reset()} the value will be passed to the
+ * <code>finalizer</code> {@code Consumer} and the next {@code #getOrCompute()}
+ * will regenerate the value.
+ */
+public final class LazyInitializable<T, E extends Exception> {
+
+    private final CheckedSupplier<T, E> supplier;
+    private final Consumer<T> primer;
+    private final Consumer<T> finalizer;
```

Comment:
how about `onReset` instead of `finalizer`?  `finalizer` makes me think of garbage collection